src/main/java/au/org/ala/names/index/BareName.java

```diff
@@ -0,0 +1,138 @@
+     *
+     * @return The new unranked scientific name
+     */
+    @Override
+    UnrankedScientificName createConcept(NameKey stageKey) {
+        return new UnrankedScientificName(this, stageKey);
+    }
+
+
+    /**
+     * Create a key for this stage.
+     *
+     * @param instanceKey
+     *
+     * @return A key that corresponds to the keys used by the concept map
+     */
+    NameKey buildStageKey(NameKey instanceKey) {
+        return instanceKey.toUnrankedNameKey();
+    }
+
+
+    /**
+     * If we have a principal, then any uncoded names are reallocated to the principal
+     *
+     * @param taxonomy The resolving taxonomy
+     * @param principal The principal scientific name
+     */
+    @Override
+    protected void reallocateDanglingConcepts(Taxonomy taxonomy, UnrankedScientificName principal) {
+        if (principal == null)
+            return;
+        List<UnrankedScientificName> reallocated = new ArrayList<>();
+        for (UnrankedScientificName name: this.getConcepts()) {
+            if (name != principal && name.getKey().isUncoded()) {
+                principal.reallocate(name, taxonomy);
+                reallocated.add(name);
+            }
+        }
+        this.removeConcepts(reallocated);
+        taxonomy.count("count.resolve.uncodedScientificName.principal");
+    }
+
+    /**
+     * Find a principal unranked name.
+     * <ul>
+     *     <li>If there is only one uncoded name, then that is it.</li>
+     *     <li>If there is only one coded scientific name, then that is it.</li>
+     *     <li>If there is more than one coded scientific name, then choose the one with a principal with the highest score. (Or the first if there are multiple ones)</li>
+     * </ul>
+     * @param taxonomy The resolving taxonomy.
+     *
+     * @return
+     */
+    @Override
+    protected UnrankedScientificName findPrincipal(Taxonomy taxonomy) {
+        List<UnrankedScientificName> names = this.getConcepts();
+
+        if (names.isEmpty())
+            return null;
+        if (names.size() == 1)
+            return names.get(0);
+        names.sort((tc1, tc2) -> tc2.getProviderScore() - tc1.getProviderScore());
+        final int cutoff = taxonomy.getAcceptedCutoff();
+        List<UnrankedScientificName> coded = names.stream().filter(sn -> !sn.getKey().isUncoded() && sn.getPrincipal() != null && sn.getPrincipalScore() > cutoff).collect(Collectors.toList());
+        if (coded.size() == 0)
+            return names.get(0);
+        if (coded.size() == 1) {
+            return coded.get(0);
+        }
+        taxonomy.report(IssueType.COLLISION, "uncodedScientificName.collision", this,coded.get(0), coded.get(1));
+        final int score = coded.stream().mapToInt(UnrankedScientificName::getPrincipalScore).max().orElse(Integer.MIN_VALUE);
+        List<UnrankedScientificName> candidates = coded.stream().filter(sn -> sn.getPrincipalScore() == score).collect(Collectors.toList());
+        if (candidates.size() > 1)
```

Comment:
To get to this point, there has to be at least one coded name with a score. If that is the case, then it will have a maximum and at least one name that matches that maximum. The .orElse could be replaced by a getAsInt() if that makes things more clear.